File: update_readme.js
const thisYear = new Date().getFullYear()
const startTimeOfThisYear = new Date(`${thisYear}-01-01T00:00:00+00:00`).getTime()
const endTimeOfThisYear = new Date(`${thisYear}-12-31T23:59:59+00:00`).getTime()
const progressOfThisYear = (Date.now() - startTimeOfThisYear) / (endTimeOfThisYear - startTimeOfThisYear)
const progressBarOfThisYear = generateProgressBar()

function generateProgressBar() {
    const progressBarCapacity = 30
    const passedProgressBarIndex = parseInt(progressOfThisYear * progressBarCapacity)
    const passedProgressBar = Array.from({length: passedProgressBarIndex}, _ => '█')
    const leftProgressBar = Array.from({length: progressBarCapacity - passedProgressBarIndex}, _ => '▁')
    return `{ ${passedProgressBar.concat(leftProgressBar).join('')} }`
}

const readme = `\
### Hi there 👋

⏳ Year progress ${progressBarOfThisYear} ${(progressOfThisYear * 100).toFixed(2)} %

⏰ Updated on ${new Date().toUTCString()} with JavaScript.

![build](https://github.com/shenxianpeng/shenxianpeng/workflows/build/badge.svg) ![Profile views](https://gpvc.arturio.dev/shenxianpeng)\
`

console.log(readme)

File: update_readme.py
import time
from datetime import datetime
from datetime import date

now = datetime.now()
curr_date = datetime.now().date()
this_year = datetime.now().year
start_of_this_year = date(this_year, 1, 1)
end_of_this_year = date(this_year, 12, 31)
progress_of_this_year = (curr_date - start_of_this_year) / (end_of_this_year - start_of_this_year)


def generate_progress_bar():
    progress_bar_capacity = 30
    passed_progress_bar_index = int(progress_of_this_year * progress_bar_capacity)
    passed_progress_bar = '█' * passed_progress_bar_index
    left_progress_bar = '▁' * (30 - passed_progress_bar_index)
    return " { " + passed_progress_bar + left_progress_bar + " } " + "%.2f" % (progress_of_this_year * 100) + " %"


bar = generate_progress_bar()
update_time = time.strftime('%Y-%m-%d %H:%M:%S')

with open("README.md", 'w', encoding='utf8') as f:
    f.write("### Hi there 👋\n\n⏳ Year progress " + bar + "\n\n⏰ Updated on " + update_time
            + " with Python.\n\n![build](https://github.com/shenxianpeng/year-progress/workflows/build/badge.svg) ![Profile views](https://gpvc.arturio.dev/shenxianpeng)")
    f.close()
